syntax = "proto3";

package TW.Harmony.Proto;
option java_package = "wallet.core.jni.proto";

// Input data necessary to create a signed transaction.
message SigningInput {
    // Chain identifier (256-bit number)
    bytes chain_id = 1;

    // Nonce (256-bit number)
    bytes nonce = 2;

    // Gas price (256-bit number)
    bytes gas_price = 3;

    // Gas limit (256-bit number)
    bytes gas_limit = 4;

    // Recipient's address.
    string to_address = 5;

    // Amount to send in wei (256-bit number)
    bytes amount = 6;

    // Optional payload
    bytes payload = 7;

    // Private key.
    bytes private_key = 8;

    // From shard ID (256-bit number)
    bytes from_shard_id = 9;

    // To Shard ID (256-bit number)
    bytes to_shard_id = 10;
}

// Transaction signing output.
message SigningOutput {
    // Signed and encoded transaction bytes.
    bytes encoded = 1;

    bytes v = 2;
    bytes r = 3;
    bytes s = 4;
}

message StakingTransactionInput {
    // StakeMsg 
    oneof stake_msg {
        DirectiveNewValidator new_validator_message = 1;
        DirectiveEditValidator edit_validator_message = 2;
        DirectiveDelegate delegate_message = 3;
        DirectiveRedelegate redelegate_message = 4;
        DirectiveUndelegate undelegate_message = 5;
    }

    // Nonce (256-bit number)
    bytes nonce = 6;

    // Gas price (256-bit number)
    bytes gas_price = 7;

    // Gas limit (256-bit number)
    bytes gas_limit = 8;    

    // Chain identifier (256-bit number)
    bytes chain_id = 9; 

    // Private key.
    bytes private_key = 10;
}

message StakingTransactionOutput {
    // Signed and encoded transaction bytes.
    bytes encoded = 1;

    bytes v = 2;
    bytes r = 3;
    bytes s = 4;
}

message Description {
    string name = 1;
    string identity = 2;
    string website = 3;
    string security_contact = 4;
    string details = 5;
}

message Decimal {
    bytes value = 1;
    bytes precision = 2;
}

message CommissionRate {
    Decimal rate = 1;
    Decimal max_rate = 2;
    Decimal max_change_rate = 3;
}

message DirectiveNewValidator {
    Description description = 1;
    CommissionRate commission = 2;
    bytes min_self_delegation = 3;
    string staking_address = 4;
    bytes pub_key = 5;
    bytes amount = 6;
}

message DirectiveEditValidator {
    Description description = 1;
    string staking_address = 2;
    Decimal commission_rate = 3;
    bytes min_self_delegation = 4;
}

message DirectiveDelegate {
    string delegator_address = 1;
    string validator_address = 2;
    bytes amount = 3;
}

message DirectiveRedelegate {
    string delegator_address = 1;
    string validator_src_address = 2;
    string validator_dst_address = 3;
    bytes amount = 4;
}

message DirectiveUndelegate {
    string delegator_address = 1;
    string validator_address = 2;
    bytes amount = 3;
}
